module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1373 @@
+ * second word of DVA[2] and the remaining 32 bits are stored in the upper 32
+ * bits of blk_fill. For most object types this is safe because we only encrypt
+ * level 0 blocks which means that the fill count will be 1. For DMU_OT_DNODE
+ * blocks the fill count is instead used to indicate the number of free dnodes
+ * beneath the bp. In either case, this number should still be smaller than
+ * UINT32_MAX so it is safe to store the IV in the top 32 bits of blk_fill,
+ * while leaving the bottom 32 bits of the fill count for the dnode code.
+ *
+ * Master key:
+ * This is the most important secret data of an encrypted dataset. It is used
+ * along with the salt to generate that actual encryption keys via HKDF. We
+ * do not use the master key to encrypt any data because there are theoretical
+ * limits on how much data can actually be safely encrypted with any encryption
+ * mode. The master key is stored encrypted on disk with the user's key. It's
+ * length is determined by the encryption algorithm. For details on how this is
+ * stored see the block comment in dsl_crypt.c
+ *
+ * Salt:
+ * Used as an input to the HKDF function, along with the master key. We use a
+ * 64 bit salt, stored unencrypted in the first word of DVA[2]. Any given salt
+ * can be used for encrypting many blocks, so we cache the current salt and the
+ * associated derived key in zio_crypt_t so we do not need to derive it again
+ * needlessly.
+ *
+ * Encryption Key:
+ * A secret binary key, generated from an HKDF function used to encrypt and
+ * decrypt data.
+ *
+ * Message Authenication Code (MAC)
+ * The MAC is an output of authenticated encryption modes suce ahes AES-GCM and
+ * AES-CCM. It's purpose is to ensure that an attacker cannot modify encrypted
+ * data on disk and return garbage to the application. Effectively, it is a
+ * checksum that can not be reproduced by an attacker. We store the MAC in the
+ * second 128 bits of blk_cksum, leaving the first 128 bits for a truncated
+ * regular checksum of the ciphertext which can be used for scrubbing.
+ *
+ *
+ * ZIL ENCRYPTION:
+ * ZIL blocks have their bp written to disk ahead of the associated data, so we
+ * cannot store encyrption paramaters there as we normally do. For these blocks
+ * the MAC is stored in the zil_chain_t header (in zc_mac) in a previously
+ * unused 8 bytes. The salt and IV are generated for the block on bp allocation.
+ * Since ZIL blocks are rewritten many times as new log records are added it is
+ * important that we do not reuse the IV with the same salt. To accomplish this
+ * we add in zc_nused from the zil_chain_t which should be incremented on each
+ * rewrite.
+ *
+ * CONSIDERATIONS FOR DEDUP:
+ * In order for dedup to work, we need to ensure that the ciphertext checksum
+ * and MAC are equivalent for equivalent plaintexts. This requires using the
+ * same IV and encryption key for equivalent blocks of plaindata. Normally,
```

Comment:
I used the word equivalent here because, in this circumstance, we wouldn't actually have the plaindata of the block that was written first in hand so we never actually make the comparison. We know the plaindata of both blocks match because the HMAC, the ciphertext checksum, and MAC all match. To be honest I'm not really sure what the correct thing to do is here from a grammar perspective.